geomstats/geometry/discrete_curves.py

```diff
@@ -1842,6 +1843,244 @@ def horizontal_path(t):
+                lenght = upper_bound - lower_bound
+                value += lenght * gs.dot(srv_1[x_min + i - 1], srv_2[y_min + j - 1])
+
+                if list_l[i] == list_k[j]:
+                    i += 1
+                    j += 1
+                elif list_l[i] < list_k[j]:
+                    i += 1
+                else :
+                    j += 1
+                lower_bound = upper_bound
+
+            value = math.pow(gamma_slope, 1 / 2) * value
+
+            return value
+
+        initial_srv = srv_function_resampled(initial_curve)
+        end_srv = srv_function_resampled(end_curve)
+
+        norm_squared_initial_srv = compute_integral_restricted(
+            initial_srv, initial_srv, 0, n_discretization, 0, n_discretization) \
+            / n_discretization
+        norm_squared_end_srv = compute_integral_restricted(
+            end_srv, end_srv, 0, n_discretization, 0, n_discretization) \
+            / n_discretization
+
+        tableau = (-1.0) * gs.ones((n_discretization + 1, n_discretization + 1))
+        tableau[0, 0] = 0.
+        gamma = {}
+        gamma[(0, 0)] = [(0, 0)]
+        for j in range(1, n_discretization + 1):
+            min_i = int(max(gs.floor(j / max_slope),
+                        n_discretization - max_slope * (n_discretization - j)))
+            max_i = int(min(j * max_slope,
+                        gs.ceil(n_discretization - (n_discretization - j)
+                            * (1 / max_slope))))
+            for i in range(min_i, max_i + 1):
+                minimum_column_index = int(max(0, i - max_slope))
+                minimum_line_index = int(max(0, j - max_slope))
+                for m in range(minimum_column_index, i):
+                    for k in range(minimum_line_index, j):
+                        if tableau[k, m] != -1 :
+                            new_value = tableau[k, m] + \
+                                compute_integral_restricted(
+                                initial_srv, end_srv, m, i, k, j)
+
+                            if tableau[j, i] < new_value :
+                                tableau[j, i] = new_value
+                                new_gamma = copy.deepcopy(gamma[(m, k)])
+                                new_gamma.append((i, j))
+                                gamma[(i, j)] = new_gamma
+
+        maximum_scalar_product = tableau[(n_discretization, n_discretization)] \
+            / n_discretization
+
+        dist_squared = norm_squared_initial_srv + \
+            norm_squared_end_srv - 2 * maximum_scalar_product
+
+        distance = gs.sqrt(dist_squared)
+
+        end_curve_reparametrized = reparametrization_curve(
+            end_curve, gamma[(n_discretization, n_discretization)])
+
+        geodesic = self.total_space.metric.geodesic(
+            initial_curve, end_curve_reparametrized)
+
+        return {"geodesic": geodesic, "distance": distance}
+
     def align(self, point, base_point, n_times=20, threshold=1e-3):
```

Comment:
Now we have 2 methods to align two curves and compute horizontal geodesics: (1) iterative horizontal projection and (2) dynamic programming. We should be able to call either of these methods in `align`, as well as in `horizontal_geodesic`. So what I suggest is:
- Rename `horizontal_geodesic` to `_iterative_horizontal_projection`
- Create a new `horizontal_geodesic` that calls either `_dynamic_programming` or `_iterative_horizontal_projection`
- Add a `method` variable to `align` and call `horizontal_geodesic(method=method)`.